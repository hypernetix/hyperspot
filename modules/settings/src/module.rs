use crate::api::rest::routes;
use crate::contract::client::SettingsApi;
use crate::domain::service::{Service, ServiceConfig};
use async_trait::async_trait;
use axum::Router;
use modkit::{DbModule, Module, ModuleCtx, OpenApiRegistry, RestfulModule};
use sea_orm_migration::MigratorTrait;
use std::any::Any;
use std::sync::Arc;
use tracing::{debug, info};

#[modkit::module(
    name = "settings",
    caps = [db, rest],
    client = crate::contract::client::SettingsApi
)]
#[derive(Default)]
pub struct Settings {
    service: arc_swap::ArcSwapOption<Service>,
}

#[async_trait]
impl Module for Settings {
    async fn init(&self, ctx: &ModuleCtx) -> anyhow::Result<()> {
        info!("Initializing Settings module");

        // Load module configuration
        // todo: we don't have one

        // Get database connection
        let db = ctx.db().ok_or_else(|| anyhow::anyhow!("DB required"))?;
        let db_conn = db.seaorm();

        // Create domain service with configuration
        let service_config = ServiceConfig::default();

        let service = Service::new(db_conn.clone(), service_config);
        self.service.store(Some(Arc::new(service.clone())));

        // Create and register the local client implementation
        // TODO: no gateways yet.
        // let api: Arc<dyn SettingsApi> = Arc::new(UsersInfoLocalClient::new(Arc::new(service)));
        // expose_users_info_client(ctx, &api)?;
        //info!("Settings API exposed to ClientHub");

        info!("Settings module initialized");
        Ok(())
    }

    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[async_trait]
impl DbModule for Settings {
    async fn migrate(&self, db: &db::DbHandle) -> anyhow::Result<()> {
        info!("Running Settings module database migrations");

        let conn = db.seaorm();
        crate::infra::storage::migrations::Migrator::up(conn, None).await?;

        info!("Settings module database migrations completed successfully");
        Ok(())
    }
}

impl RestfulModule for Settings {
    fn register_rest(
        &self,
        ctx: &ModuleCtx,
        router: Router,
        openapi: &dyn OpenApiRegistry,
    ) -> anyhow::Result<Router> {
        info!("Registering settings REST routes");

        let service = self
            .service
            .load()
            .as_ref()
            .ok_or_else(|| anyhow::anyhow!("Service not initialized"))?
            .clone();

        let router = routes::register_routes(router, openapi, service)?;

        info!("Settings REST routes registered successfully");
        Ok(router)
    }
}
