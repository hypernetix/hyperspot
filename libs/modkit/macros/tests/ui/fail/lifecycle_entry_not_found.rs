use modkit_macros::module;
use tokio_util::sync::CancellationToken;
use anyhow::Result;

#[module(name="x", capabilities=[stateful], lifecycle(entry="serve", await_ready))]
pub struct X;

impl X {
    // Wrong signature: missing ReadySignal parameter â†’ the generated call won't match.
    async fn serve(&self, _cancel: CancellationToken) -> Result<()> { Ok(()) }
}

#[async_trait::async_trait]
impl modkit::Module for X {
    async fn init(&self, _ctx: &modkit::ModuleCtx) -> anyhow::Result<()> { Ok(()) }
    fn as_any(&self) -> &dyn core::any::Any { self }
}

fn main() {}
