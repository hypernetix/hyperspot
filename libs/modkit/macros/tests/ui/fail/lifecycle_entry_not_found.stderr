error[E0277]: the trait bound `X: Default` is not satisfied
 --> tests/ui/fail/lifecycle_entry_not_found.rs:6:12
  |
6 | pub struct X;
  |            ^ the trait `Default` is not implemented for `X`
  |
help: consider annotating `X` with `#[derive(Default)]`
  |
6 + #[derive(Default)]
7 | pub struct X;
  |

error[E0061]: this method takes 1 argument but 2 arguments were supplied
  --> tests/ui/fail/lifecycle_entry_not_found.rs:5:1
   |
5  | #[module(name="x", capabilities=[stateful], lifecycle(entry="serve", await_ready))]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unexpected argument #2 of type `ReadySignal`
   |
note: method defined here
  --> tests/ui/fail/lifecycle_entry_not_found.rs:10:14
   |
10 |     async fn serve(&self, _cancel: CancellationToken) -> Result<()> { Ok(()) }
   |              ^^^^^
   = note: this error originates in the attribute macro `module` (in Nightly builds, run with -Z macro-backtrace for more info)
